File: DSA/LC/211_design_add_and_search_words_data_structure.cpp
#include <vector>
#include <string>
class TrieNode
{
public:
    std::vector<TrieNode *> children;
    bool word;

    TrieNode() : children(26, nullptr), word(false) {}
};

class WordDictionary
{
public:
    TrieNode *root;

    WordDictionary() : root(new TrieNode()) {}

    void addWord(std::string word)
    {
        TrieNode *cur = root;
        for (char c : word)
        {
            if (cur->children[c - 'a'] == nullptr)
            {
                cur->children[c - 'a'] = new TrieNode();
            }
            cur = cur->children[c - 'a'];
        }
        cur->word = true;
    }

    bool search(std::string word)
    {
        return dfs(word, 0, root);
    }

private:
    bool dfs(std::string word, int j, TrieNode *root)
    {
        TrieNode *cur = root;

        for (int i = j; i < word.size(); i++)
        {
            char c = word[i];
            if (c == '.')
            {
                for (TrieNode *child : cur->children)
                {
                    if (child != nullptr && dfs(word, i + 1, child))
                    {
                        return true;
                    }
                }
                return false;
            }
            else
            {
                if (cur->children[c - 'a'] == nullptr)
                {
                    return false;
                }
                cur = cur->children[c - 'a'];
            }
        }
        return cur->word;
    }
};

File: DSA/LC/211_design_add_and_search_words_data_structure.py
class TrieNode:
    def __init__(self):
        self.word = False
        self.children = {}


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        curr = self.root

        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()

            curr = curr.children[c]

        curr.word = True

    def search(self, word: str) -> bool:

        def dfs(root, i):

            if i == len(word):
                return root.word
            if word[i] == '.':
                for k in root.children:
                    if dfs(root.children[k], i + 1):
                        return True

            elif word[i] in root.children:
                return dfs(root.children[word[i]], i + 1)

            return False
        return dfs(self.root, 0)
